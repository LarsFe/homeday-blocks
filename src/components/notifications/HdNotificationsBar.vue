<template>
  <div
    :class="{
      'notifications-bar': true,
      [`notifications-bar--${type}`]: true,
      isVisible: visible,
      isCompact: compact,
    }"
  >
    <div class="notifications-bar__icon">
      <HdIcon
        :src="icon"
        class="notifications-bar__icon__image"
      />
    </div>
    <div class="notifications-bar__content">
      <slot>
        {{ message }}
      </slot>
    </div>
  </div>
</template>

<script>
import HdIcon from 'homeday-blocks/src/components/HdIcon.vue';
import {
  verifiedIcon,
  reportIcon,
  warningIcon,
  bellIcon,
} from 'homeday-blocks/src/assets/small-icons';

export default {
  name: 'HdNotificationsBar',
  components: {
    HdIcon,
  },
  props: {
    type: {
      type: String,
      default: 'notification',
      validator(type) {
        const allowedTypes = [
          'notification',
          'success',
          'error',
          'warning',
        ];

        return allowedTypes.some(allowedType => allowedType === type);
      },
    },
    message: {
      type: String,
      default: '',
    },
    visible: {
      type: Boolean,
      default: true,
    },
    customIcon: {
      type: String,
      default: '',
    },
    compact: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    icon() {
      if (this.customIcon) {
        return this.customIcon;
      }

      if (this.type === 'success') {
        return verifiedIcon;
      }

      if (this.type === 'error') {
        return reportIcon;
      }

      if (this.type === 'warning') {
        return warningIcon;
      }

      return bellIcon;
    },
  },
};
</script>

<style lang="scss">
@import 'homeday-blocks/src/styles/mixins.scss';

.notifications-bar {
  position: absolute;
  top: 0;
  right: 0;
  left: 0;
  display: flex;
  justify-content: center;
  opacity: 0;
  padding: $stack-m $inline-s;
  transition: opacity ($time-s * 2) ease-in-out;
  z-index: 10;

  &.isVisible {
    opacity: 1;
  }

  &.isCompact {
    padding: $stack-s $inline-s;
  }

  &--enter-active,
  &--leave-active {
    transition: all ($time-s * 2) ease-in-out;
  }

  &--enter,
  &--leave-to {
    transform: translateY(-100%);
  }

  &--notification {
    background-color: $tertiary-color;
    color: $white;
  }

  &--success {
    background-color:$success-color;
    color: $white;
  }

  &--error {
    background-color:#E00016;
    color: $white;
  }

  &--warning {
    background-color:#FFAB00;
    color: $primary-color;
  }

  &__icon {
    align-items: center;
    display: flex;
    height: 24px;
    justify-content: center;
    margin-right: $inline-s;
    width: 24px;

    path {
      fill: currentColor;
    }
  }

  &__content {
    font-size: 14px;
    font-weight: 600;
    line-height: 24px;

    @media (min-width: $break-tablet) {
      font-size: 18px;
    }
  }
}
</style>
