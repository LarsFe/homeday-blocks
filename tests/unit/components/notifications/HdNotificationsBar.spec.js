import { mount } from '@vue/test-utils';
import HdNotificationsBar from '@/components/notifications/HdNotificationsBar.vue';

const MESSAGE = 'Hello world!';
const SLOT = '<p>Hello world!, <a href="#">This is a link...</a></p>';
const CONTENT_SELECTOR = '.notifications-bar__content';
const ICON_FAKE_PATH = '/foo/bar/icon.svg';

const COMPACT_CLASS = 'notifications-bar--compact';
const VISIBLE_CLASS = 'notifications-bar--visible';

describe('HdNotificationsBar', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(HdNotificationsBar, {
      propsData: {
        message: MESSAGE,
      },
    });
  });

  it('renders component', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders a message correctly', () => {
    expect(wrapper.find(CONTENT_SELECTOR).text()).toBe(MESSAGE);
  });

  it('renders a slot correctly', () => {
    wrapper = mount(HdNotificationsBar, {
      slots: {
        default: SLOT,
      },
    });

    expect(wrapper.find(CONTENT_SELECTOR).element.innerHTML).toBe(SLOT);
  });

  it('supports custom icon', () => {
    wrapper.setProps({ customIcon: ICON_FAKE_PATH });
    expect(wrapper.vm.icon).toBe(ICON_FAKE_PATH);
  });

  it('can be compact', () => {
    wrapper.setProps({ compact: true });
    expect(wrapper.classes().includes(COMPACT_CLASS)).toBe(true);
  });

  it('supports hiding the bar', () => {
    expect(wrapper.classes().includes(VISIBLE_CLASS)).toBe(true);
    wrapper.setProps({
      visible: false,
    });
    expect(wrapper.classes().includes(VISIBLE_CLASS)).toBe(false);
  });
});
