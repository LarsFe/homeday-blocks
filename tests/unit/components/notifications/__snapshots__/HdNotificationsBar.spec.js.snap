// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdNotificationsBar renders component 1`] = `
<div class="notifications-bar notifications-bar--notification notifications-bar--visible">
  <div class="notifications-bar__icon">
    <!---->
  </div>
  <div class="notifications-bar__content">
    Hello world!
  </div>
</div>
`;
